fh-sync.d.ts

```diff
@@ -0,0 +1,412 @@
+// Type definitions for fh-sync
+// Project: https://github.com/feedhenry/fh-sync
+// Maintainer feedhenry-dev@redhat.com
+
+declare module SyncCloud {
+
+      /**
+       * Backoff strategies that can be passed to workers
+       */
+      type WorkerBackoffStrategy = 'fib'|'exp'
+
+      /**
+       * Valid actions (the "fn" param) that can be passed to sync.invoke
+       */
+      type InvokeAction = 'sync'|'syncRecords'|'listCollisions'|'removeCollision'
+
+      /**
+       * Options that can be passed to sync.invoke
+       */
+      interface InvokeOptions {
+          fn: InvokeAction
+
+          // TODO: we should define this in more detail
+          [key: string]: any
+      }
+
+      /**
+       * Interfaces that describe how sync responses should be structured
+       */
+      namespace HandlerResults {
+          interface Create {
+              uid: string
+              data: Object
+          }
+
+          interface Read {
+              [key: string]: any
+          }
+
+          interface Update {
+              [key: string]: any
+          }
+
+          interface Delete {
+              [key: string]: any
+          }
+
+          interface List {
+              [uid: string]: Object
+          }
+      }
+
+      /**
+       * Unique callback structures for sync handlers.
+       *
+       * Can be used by other modules to create pre-built sync compliant handlers.
+       */
+      namespace HandlerFunctions {
+          type Create = (dataset: string, data: Object, metaData: any, done: StandardCb<HandlerResults.Create>) => void
+          type Read = (dataset: string, uid: string, metaData: any, done: StandardCb<HandlerResults.Read>) => void
+          type Update = (dataset: string, uid: string, metaData: any, done: StandardCb<HandlerResults.Update>) => void
+          type Delete = (dataset: string, queryParams: any, metaData: any, done: StandardCb<HandlerResults.Delete>) => void
+          type List = (dataset: string, queryParams: any, metaData: any, done: StandardCb<HandlerResults.List>) => void
+          type Collision = (datasetId: string, hash: string, timestamp: number, uid: string, pre: Object, post: Object, metaData: any, callback: StandardCb<any>) => void
+          type ListCollisions = (datasetId: string, metaData: any, callback: StandardCb<{ [hash: string]: Object }>) => void
+          type RemoveCollision = (datasetId: string, collision_hash: string, metaData: any, callback: StandardCb<any>) => void
+          type Interceptor = (datasetId: string, interceptorParams: SyncInterceptParams, callback: NoRespCb) => void
+          type Hash = (data: any) => string
```

Comment:
There are two types of hash function callbacks. The record one gets `(datasetId: string, record: any)` and the global hash one gets `Array<any>`

Reply:
Oooh, well spotted! Will fix that.

Reply:
@MikeyBurkman I think the `(datasetId: string, record: any)` is no longer valid. I have some TS code now like this that is working with these types:

```ts
const hashHandler: sync.HandlerFunctions.Hash = (t: Item) => {
  return `${t.name}${t.lastUpdated}`
}
```

Our documentation states the signature is:

```js
$fh.sync.setRecordHashFn(dataset_id, function generateHash(dataset_id,
record){});
```

Based on this [code](https://github.com/feedhenry/fh-sync/blob/master/lib/hashProvider.js#L23) the implementation I have is correct, so the docs might need updating.

Take a look at the code I'll push shortly